src/header/common/range.rs

```diff
@@ -0,0 +1,199 @@
+/// > semantics to request transfer of only one or more subranges of the
+/// > selected representation data, rather than the entire selected
+/// > representation data.
+///
+/// > Range = byte-ranges-specifier / other-ranges-specifier
+/// > other-ranges-specifier = other-range-unit "=" other-range-set
+/// > other-range-set = 1*VCHAR
+///
+/// # Example values
+/// * `bytes=1000-`
+/// * `bytes=-2000`
+/// * `bytes=0-1,30-40`
+/// * `custom_unit=0-123`
+///
+/// Spec: https://tools.ietf.org/html/rfc7233#section-3.1
+#[derive(PartialEq, Clone, Debug)]
+pub struct Range {
+    /// Unit of the Range i.e. bytes
+    pub unit: RangeUnit,
+    /// Set of ranges as defined in the HTTP spec
+    pub ranges: Vec<RangeSpec>,
+}
+
+/// Each 'Range' header can contain one or more RangeSpecs.
+/// Each RangeSpec defines a range of units to fetch
+#[derive(PartialEq, Clone, Debug)]
+pub enum RangeSpec {
+    /// Get all bytes between x and y ("x-y")
+    FromTo(u64, u64),
+    /// Get all bytes starting from x ("x-")
+    AllFrom(u64),
+    /// Get last x bytes ("-x")
+    Last(u64)
+}
+
+impl Range {
+    /// Get the most common byte range header ("bytes=from-to")
+    pub fn bytes(from: u64, to: u64) -> Range {
+        Range {
+            unit: RangeUnit::Bytes,
+            ranges: vec![RangeSpec::FromTo(from, to)],
+        }
+    }
+}
+
+
+impl fmt::Display for RangeSpec {
+    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
+        match *self {
+            RangeSpec::FromTo(from, to) => write!(fmt, "{}-{}", from, to),
+            RangeSpec::Last(pos) => write!(fmt, "-{}", pos),
+            RangeSpec::AllFrom(pos) => write!(fmt, "{}-", pos),
+        }
+    }
+}
+
+
+impl fmt::Display for Range {
+    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
+        try!(write!(fmt, "{}=", self.unit));
+
+        let mut ranges = self.ranges.iter();
+        match ranges.next() {
+            Some(range) => try!(write!(fmt, "{}", range)),
+            None => return Ok(())
+        }
+        for range in ranges {
+            try!(write!(fmt, ",{}", range));
+        }
+        Ok(())
+    }
+}
+
+impl FromStr for Range {
+    type Err = ();
+
+    fn from_str(s: &str) -> Result<Range, ()> {
+        let mut iter = s.split("=");
+
+        let (unit, ranges) = match (iter.next(), iter.next()) {
+            (Some(unit), Some(ranges)) => (unit, ranges),
+            _ => return Err(())
+        };
+
+        if let Ok(unit) = RangeUnit::from_str(unit) {
+            let mut valid_ranges: Vec<RangeSpec> = vec![];
+            for range in ranges.split(",") {
+                valid_ranges.push(try!(FromStr::from_str(range)));
+            }
+            return Ok(Range{unit: unit, ranges: valid_ranges});
+        }
+
+        Err(())
+    }
+}
+
+impl FromStr for RangeSpec {
+    type Err = ();
+
+    fn from_str(s: &str) -> Result<RangeSpec, ()> {
+        let mut parts = s.split("-");
```

Comment:
You should probably use [`splitn(2)`](https://doc.rust-lang.org/std/primitive.str.html#method.splitn) here. Imagine a malformed `RangeSpec` like `200-300-foobar`. It would pass as valid.